[IGNORE] Add authorization doc (#1481)
perses/perses: docs/authorization.md

# Authorization / Permission

*WIP: The authorization is still not implemented in Perses*

Perses will use a Role-based access control (RBAC) for regulating access to resources based on the role of the user.
The RBAC API is based on four kinds of resource: GlobalRole, Role, GlobalRoleBinding and RoleBinding.
Perses RBAC implementation is highly inspired by K8S RBAC implementation.

## Role and GlobalRole

> An RBAC *Role* or *GlobalRole* contains a set of permissions (rules). Permissions are purely additive (there are no "deny" permissions).

Reviewer: ```suggestion
An RBAC `Role` or `GlobalRole` contains a set of permissions (rules). Permissions are purely additive (there are no "deny" permissions).

> A *Role* set permission within a particular project, when you create a *Role* you need to specify the project it belongs in.

Reviewer: ```suggestion
A `Role` set permission within a particular project, when you create a *Role* you need to specify the project it belongs in.
```

> *GlobalRole*, by contrast, is not limited to a project scope.

Reviewer: ```suggestion
`GlobalRole`, by contrast, is not limited to a project scope.
```

### Role example

Here is an example of Role in "MySuperProject" project that can be used to grant edit access to dashboards:
```yaml
kind: Role
metadata:
  name: dashboard-editor
  project: MySuperProject
spec:
  permissions:
    - action: edit
      scopes: ["Dashboard"]
```

### Global Role example

> A *GlobalRole* can be used to grant the same permissions as a Role. However, because *GlobalRole* are global, you can also use them to grant access to:

Reviewer: ```suggestion
A `GlobalRole` can be used to grant the same permissions as a Role. However, because `GlobalRole` are global, you can also use them to grant access to:
```

- global resources (like Global Datasources, Global Variables, Users, ...)
- project resources (like Dashboards) across all projects

Here is an example of a *GlobalRole* that can be used to grant edit access to variables in all projects:
```yaml
kind: GlobalRole
metadata:
  name: variable-editor
spec:
  permissions:
    - action: edit
      scopes: ["Variable"]
```

## RoleBinding and GlobalRoleBinding

A role binding grants the permissions defined in a role to a user or set of users. 
It holds a list of subjects (users or teams) and a reference to the role being granted. A *RoleBinding* grants permissions within a specific namespace whereas a *GlobalRoleBinding* grants that access global-wide.

> A *RoleBinding* may reference any *Role* in the same namespace. Similarly, a *ClusterRoleBinding* can reference any *ClusterRole*.

Reviewer 2: ```suggestion
A *RoleBinding* may reference any *Role* in the same project. Similarly, a *GlobalRoleBinding* can reference any *GlobalRole*.
```

Reviewer 2: Are you sure it's namespace and not project here?

- A binding to a different role is a fundamentally different binding. Requiring a binding to be deleted/recreated in order to change the roleRef ensures the full list of subjects in the binding is intended to be granted the new role (as opposed to enabling or accidentally modifying only the roleRef without verifying all of the existing subjects should be given the new role's permissions).

## Referring to resources

In Perses API, resources are identified and accessed using a string, corresponding to the name in the metadata. You can also refer to all resource using the wildcard `*` character.
Here is an example for granting edit permissions to all resources in all projects:
```yaml
kind: GlobalRole
metadata:
  name: admin-editor
spec:
  permissions:
    - action: edit
      scopes: ["*"]
```

## RBAC Synchro

Roles and RoleBindings of an user are stored in the user's JWT.
If Perses is deployed with multiple instances, Perses RBAC roles and role bindings cache need to be synchronized/replicated between all instances.
To do that there are multiple mechanisms:
- cache is refreshed every X minutes

> - cache is refreshed if roles and role bindings retrieve from the user's JWT are different from the cache

Reviewer: I would say it is refreshed only if the token is saying the user has more right than what is contained in the cache.

> - cache is refresh when a new role is created, edited or deleted

Reviewer 2: ```suggestion
- cache is refreshed when a new role is created, edited or deleted
```

Reviewer 2: I think the correct one is to put refresh**ed** on all rows

> - cache is refresh when a new rolebinding is created, edited or deleted

Reviewer 2: ```suggestion
- cache is refreshed when a new rolebinding is created, edited or deleted
```
